musig2test.c

```diff
@@ -40,114 +26,204 @@ int main(void) {
+        }
     }
     printf("* %d Signers initialized.\n", N);
     printf("--------------------------------------------------------------------------- \n");
-    /****************************************************************/
-
-
-
-    /* Collect signers **********************************************/
-    printf("\n-------- Collect Signers -------------------------------------------------- \n");
-    secp256k1_pubkey X_LIST[N];         // Signers' public key list
-    unsigned char* ser_xonly_XLIST[N];  // The list of signers' serialized public key list
-    unsigned char L[XONLY_BYTES*N];     // Concatenated ser_xonly_XLIST
-
-    /* Store full-size public keys of signers in X_LIST.
-     * Get x_only version of public keys.
-     * Store serialized x_only public keys in ser_xonly_XLIST (to compute L).
-     * Concat the elements of ser_xonly_XLIST in L (to compute a). */
-    for(i=0; i<N; i++){
-        secp256k1_xonly_pubkey xonly_TEMP;
-        ser_xonly_XLIST[i] = malloc(XONLY_BYTES);
-        if (secp256k1_keypair_pub(ctx, &X_LIST[i], signer_LIST[i]->keypair) &&
-            secp256k1_keypair_xonly_pub(ctx, &xonly_TEMP, NULL, signer_LIST[i]->keypair))
-        {
-            secp256k1_xonly_pubkey_serialize(ctx, ser_xonly_XLIST[i], &xonly_TEMP);
-            printf("* X%d: ", i+1);
-            print_hex(ser_xonly_XLIST[i], XONLY_BYTES);
-            memcpy( &L[i*XONLY_BYTES], ser_xonly_XLIST[i], XONLY_BYTES);
+
+
+    printf("\n*************************************************************************** \n");
+    printf("* STATE 1  \n");
+    printf("*************************************************************************** \n");
+
+    /**** Aggregate the public keys and batch commitments for each signer ****/
+    printf("\n-------- Aggregate PK and R ----------------------------------------------- \n");
+    for (i=0; i<N; i++) {
+        mcs_list[i]->mc->state = 0;
+        musig2_aggregate_pubkey(mcs_list[i]->mc, pk_list);
+        musig2_agg_R(mcs_list[i], batch_list);
+    }
+    printf("* Public key aggregated.\n");
+    printf("* Commitments aggregated.\n");
+    printf("--------------------------------------------------------------------------- \n");
+
+
+    /**** Signature ****/
+    printf("\n-------- Partial Signatures ----------------------------------------------- \n");
+    for (i=0; i<N; i++) {
+        /* Generate the partial signatures */
+        musig2_sign(mcs_list[i], MSG_1, TAG_1);
+        mcs_list[i]->mc->state++;
+
+        printf(" S%d: ", i+1);
+        print_hex(mcs_list[i]->parsig, SCALAR_BYTES);
+
+        /* Collect the partial signatures in parsig_list */
+        parsig_list[i] = malloc(SCALAR_BYTES);
+        memcpy(&parsig_list[i][0], mcs_list[i]->parsig, SCALAR_BYTES);
+
+        /* Check whether all aggregated R is same */
+        for (j=0; j<i; j++){
+            if (secp256k1_ec_pubkey_cmp(ctx, &mcs_list[i]->mc->R, &mcs_list[j]->mc->R )  != 0){
+                return -1;
+            }
```

Comment:
This can be done in the aggregate function, but in this case, we need to collect the `mcs_list[i]->mc->R` in a local array and pass that as a parameter to the aggregate or init_aggregator function.  I can change it as you decide?

Reply:
You could define the partial_signature list to be not only the scalar, but also the value R. In this way, when the aggregator receives all signatures, it can verify that all R values are equal